corehq/motech/repeaters/tasks.py

```diff
@@ -72,65 +64,112 @@ def delete_old_request_logs():
-        metrics_counter("commcare.repeaters.check.locked_out", tags={'partition': partition})
+        metrics_counter("commcare.repeaters.check.locked_out")
         return
 
     try:
         with metrics_histogram_timer(
             "commcare.repeaters.check.processing",
             timing_buckets=_check_repeaters_buckets,
         ):
-            for record in RepeatRecord.objects.iter_partition(
-                    start, partition, CHECK_REPEATERS_PARTITION_COUNT):
-                if not _soft_assert(
-                    datetime.utcnow() < twentythree_hours_later,
-                    "I've been iterating repeat records for 23 hours. I quit!"
-                ):
-                    break
-
+            for repeater in iter_ready_repeaters():
                 metrics_counter("commcare.repeaters.check.attempt_forward")
-                record.attempt_forward_now(is_retry=True)
-            else:
-                iterating_time = datetime.utcnow() - start
-                _soft_assert(
-                    iterating_time < timedelta(hours=6),
-                    f"It took {iterating_time} to iterate repeat records."
-                )
+                process_repeater.delay(repeater.domain, repeater.repeater_id)
+
+                if datetime.utcnow() > twentythree_hours_later:
+                    # Break after process_repeater() to avoid a stale lock
+                    break
     finally:
         check_repeater_lock.release()
 
 
+def iter_ready_repeaters():
+    """
+    Cycles through repeaters (repeatedly ;) ) until there are no more
+    repeat records ready to be sent.
+    """
+    while True:
+        yielded = False
+        for repeater in Repeater.objects.all_ready():
+            if not repeater.domain_can_forward:
+                continue
+
+            if rate_limit_repeater(repeater.domain):
+                repeater.rate_limit()
+                continue
+
+            yielded = True
+            yield repeater
+
+        if not yielded:
+            # No repeaters are ready, or they are rate limited, or their
+            # domains can't forward or are paused.
+            return
+
+
+@task(queue=settings.CELERY_REPEAT_RECORD_QUEUE)
+def process_repeater(domain, repeater_id):
+
+    def is_retry(repeat_record):
+        return repeat_record.state == State.Fail
+
+    lock_key = f'process_repeater_{repeater_id}'
+    repeater_lock = get_redis_lock(lock_key, timeout=23 * 60 * 60, name=lock_key)
+    if not repeater_lock.acquire(blocking=False):
+        return
+
+    try:
+        repeater = Repeater.objects.get(domain=domain, id=repeater_id)
+        repeat_records = repeater.repeat_records_ready[:repeater.num_workers]
+        header_tasks = [
+            retry_process_repeat_record.s(rr.id, rr.domain)
+            if is_retry(rr)
+            else process_repeat_record.s(rr.id, rr.domain)
+            for rr in repeat_records
+        ]
+        chord(header_tasks)(update_repeater.s(repeater.repeater_id))
```

Comment:
This doesn't wait for the tasks to complete right? Just wondering about the lock and what it's purpose is since it's not actually locking the sending of the records. I'm not sure locking is needed at all.

Reply:
I think my first take was actually correct, and my tests were confusing me. 7c4e3cf9acf

Reply:
(My first take was not correct. I'm not sure why this isn't working with Redis, because I'm sure it should. I have dropped locking repeaters for now. I imagine that not locking is a problem especially when a remote endpoint is timing out. Locking using Postgres, either on the Repeater model, or using a lock table, might be an alternative.)

Reply:
> Just wondering about the lock and what it's purpose is

I discovered an important purpose for the lock. If `prefetch_multiplier` is not set to a small number (it is "1" on Production, and not set on Staging) then Celery will keep iterating repeaters, and queue up as many tasks as it thinks is reasonable, and it is not reasonable. On Staging (which has concurrency set to 4) it queued up 4,814,913 tasks. This blocked the queue for all but the domain with the most repeat records. Pausing data forwarding for the offending domain had no effect because the tasks were already queued, and so Celery continued to try to process its repeat records. (Purging the tasks in the queue fixed the problem.)

(Edit: I moved this comment from the wrong thread.)

Reply:
2bb7c595054